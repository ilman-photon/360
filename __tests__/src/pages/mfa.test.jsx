import { render, fireEvent, waitFor } from "@testing-library/react";
import MfaPage from '../../../src/pages/patient/mfa';
import "@testing-library/jest-dom";
import MockAdapter from "axios-mock-adapter";
import axios from "axios";
import Cookies from "universal-cookie";

jest.mock("universal-cookie", () => {
    class MockCookies {
        static result = {};
        get(param) {
            if (param === "username") return "user1@photon.com"
            if (param === "ip") return "10.10.10.10"
            return MockCookies.result;
        }
        remove() {
            return jest.fn();
        }
        set() {
            return jest.fn();
        }
    }
    return MockCookies;
});

describe("Multi-Factor Authentication", () => {
    const mock = new MockAdapter(axios);
    let container
    beforeEach(async () => {
        Cookies.result = { mfa: true };

        const userData = {
            "communicationMethod": {
                "email": "user1@photon.com",
                "phone": "9998887772"
            },
            "ResponseCode": 4000,
            "ResponseType": "success",
        }

        mock.onPost(`/ecp/patient/mfa/getUserData`).reply(200, userData);

        container = render(<MfaPage />)
        await waitFor(() => container.getByText("communicationMethodTitle"));

    });

    test("is set MFA page render", () => {
        const title = container.getByText("setMFATitle");
        expect("setMFATitle").toEqual(title.textContent);
    });

    test("is confirm button clicked", async () => {
        const data = {
            "mfaCode": 660927,
            "ResponseCode": 4000,
            "ResponseType": "success"
        }
        mock.onPost(`/ecp/patient/mfa/sendotp`).reply(200, data);
        const confirmButton = container.getByRole("button", { name: /confrimBtn/i });
        fireEvent.click(confirmButton)

        await waitFor(() => container.getByText("mfaTitle"))

        const title = container.getByText("mfaTitle");
        expect("mfaTitle").toEqual(title.textContent);
    });

    test("is onResendCode button clicked", async () => {
        const data = {
            "mfaCode": 123456,
            "ResponseCode": 4000,
            "ResponseType": "success"
        }
        mock.onPost(`/ecp/patient/mfa/sendotp`).reply(200, data);
        const confirmButton = container.getByRole("button", { name: /confrimBtn/i });
        fireEvent.click(confirmButton)

        await waitFor(() => container.getByRole("button", { name: /resendCodeBtn/i }))

        const resendCodeButton = container.getByRole("button", { name: /resendCodeBtn/i });
        fireEvent.click(resendCodeButton)

        await waitFor(() => container.getByText("mfaTitle"))

        const title = container.getByText("mfaTitle");
        expect("mfaTitle").toEqual(title.textContent);
    })

    test("is submit button clicked", async () => {
        const data = {
            "ResponseCode": 4000,
            "ResponseType": "success"
        }
        mock.onPost(`/ecp/patient/mfa/verifyotp`).reply(200, data);
        const confirmButton = container.getByRole("button", { name: /confrimBtn/i });
        fireEvent.click(confirmButton)

        await waitFor(() => container.getByRole("button", { name: /submitBtn/i }))

        const mfaField = container.getByLabelText("mfaLabel");
        fireEvent.change(mfaField, { target: { value: "123456" } });

        const submitButton = container.getByRole("button", { name: /submitBtn/i });
        fireEvent.click(submitButton)

        await waitFor(() => container.getByText("mfaTitle"))

        const title = container.getByText("mfaTitle");
        expect("mfaTitle").toEqual(title.textContent);
    })

    test("is submit button clicked with remember me", async () => {
        const data = {
            "ResponseCode": 4000,
            "ResponseType": "success",
            "mfaAccessToken": "eyJhbGciOiJIUzI1NiJ9.eyJ1c2VybmFtZSI6InVzZXIzQHBob3Rvbi5jb20iLCJzdWIiOiJ1c2VyM0BwaG90b24uY29tIiwianRpIjoiMDBkMDhkYmMtMWQwYy00MTRlLTliN2EtNDdhNmI1MTY5MjYyIiwiaWF0IjoxNjYwODE3MjQwLCJleHAiOjE2Njg1OTMyNDB9.ALpmECst225hVk6X1d6QZutKN-Ykyl5pZ3K8LlIBmE4"
        }
        mock.onPost(`/ecp/patient/mfa/verifyotp`).reply(200, data);
        const confirmButton = container.getByRole("button", { name: /confrimBtn/i });
        fireEvent.click(confirmButton)

        await waitFor(() => container.getByRole("button", { name: /submitBtn/i }))

        const remeberMe = container.getByText("rememberMeLabel");
        fireEvent.click(remeberMe);

        const mfaField = container.getByLabelText("mfaLabel");
        fireEvent.change(mfaField, { target: { value: "123456" } });

        const submitButton = container.getByRole("button", { name: /submitBtn/i });
        fireEvent.click(submitButton)

        await waitFor(() => container.getByText("mfaTitle"))

        const title = container.getByText("mfaTitle");
        expect("mfaTitle").toEqual(title.textContent);
    })

    test("is submit button clicked with new mfa code", async () => {
        const data = {
            "mfaCode": 111111,
            "ResponseCode": 4000,
            "ResponseType": "success"
        }
        mock.onPost(`/ecp/patient/mfa/sendotp`).reply(200, data);
        const submitData = {
            "ResponseCode": 4000,
            "ResponseType": "success"
        }
        mock.onPost(`/ecp/patient/mfa/verifyotp`).reply(200, submitData);
        const confirmButton = container.getByRole("button", { name: /confrimBtn/i });
        fireEvent.click(confirmButton)

        await waitFor(() => container.getByRole("button", { name: /resendCodeBtn/i }))

        const resendCodeButton = container.getByRole("button", { name: /resendCodeBtn/i });
        fireEvent.click(resendCodeButton)

        await waitFor(() => container.getByRole("button", { name: /submitBtn/i }))

        const mfaField = container.getByLabelText("mfaLabel");
        fireEvent.change(mfaField, { target: { value: "111111" } });

        const submitButton = container.getByRole("button", { name: /submitBtn/i });
        fireEvent.click(submitButton)

        const title = container.getByText("mfaTitle");
        expect("mfaTitle").toEqual(title.textContent);
    })

    test("is submit button clicked wrong mfa", async () => {
        const data = {
            "mfaCode": 123456,
            "ResponseCode": 4000,
            "ResponseType": "success"
        }
        mock.onPost(`/ecp/patient/mfa/sendotp`).reply(200, data);
        const submitData = {
            "ResponseCode": 4002,
            "ResponseType": "Incorrect Code. Please try again"
        }
        mock.onPost(`/ecp/patient/mfa/verifyotp`).reply(200, submitData);
        const confirmButton = container.getByRole("button", { name: /confrimBtn/i });
        fireEvent.click(confirmButton)

        await waitFor(() => container.getByRole("button", { name: /submitBtn/i }))

        const mfaField = container.getByLabelText("mfaLabel");
        fireEvent.change(mfaField, { target: { value: "999999" } });

        const submitButton = container.getByRole("button", { name: /submitBtn/i });
        fireEvent.click(submitButton)

        await waitFor(() => container.getByText("mfaFailedTitle"))

        const errorTitle = container.getByText("mfaFailedTitle");
        expect("mfaFailedTitle").toEqual(errorTitle.textContent);
    })

    test("is submit button clicked locked", async () => {
        const data = {
            "mfaCode": 123456,
            "ResponseCode": 4000,
            "ResponseType": "success"
        }
        mock.onPost(`/ecp/patient/mfa/sendotp`).reply(200, data);
        const submitData = {
            "ResponseCode": 4003,
            "ResponseType": "Locked"
        }
        mock.onPost(`/ecp/patient/mfa/verifyotp`).reply(200, submitData);
        const confirmButton = container.getByRole("button", { name: /confrimBtn/i });
        fireEvent.click(confirmButton)

        await waitFor(() => container.getByRole("button", { name: /submitBtn/i }))

        const mfaField = container.getByLabelText("mfaLabel");
        fireEvent.change(mfaField, { target: { value: "999999" } });

        const submitButton = container.getByRole("button", { name: /submitBtn/i });
        fireEvent.click(submitButton)

        await waitFor(() => container.getByText("mfaLockTitle"))

        const errorTitle = container.getByText("mfaLockTitle");
        expect("mfaLockTitle").toEqual(errorTitle.textContent);
    })

    test("is onResendCode button catch", async () => {
        let data = {
            "mfaCode": 123456,
            "ResponseCode": 4000,
            "ResponseType": "success"
        }
        mock.onPost(`/ecp/patient/mfa/sendotp`).reply(200, data);
        const confirmButton = container.getByRole("button", { name: /confrimBtn/i });
        fireEvent.click(confirmButton)

        data = {
            "ResponseCode": 4001,
            "ResponseType": "Please try setting up MultiFactor Authentication after 30 minutes"
        }
        mock.onPost(`/ecp/patient/mfa/sendotp`).reply(400, data);

        await waitFor(() => container.getByRole("button", { name: /resendCodeBtn/i }))

        const resendCodeButton = container.getByRole("button", { name: /resendCodeBtn/i });
        fireEvent.click(resendCodeButton)

        await waitFor(() => container.getByText("mfaTitle"))

        const title = container.getByText("mfaTitle");
        expect("mfaTitle").toEqual(title.textContent);
    })

    test("is confirm button clicked catch", async () => {
        const data = {
            "ResponseCode": 4004,
            "ResponseType": "Please try setting up MultiFactor Authentication after 30 minutes"
        }
        mock.onPost(`/ecp/patient/mfa/sendotp`).reply(400, data);
        const confirmButton = container.getByRole("button", { name: /confrimBtn/i });
        fireEvent.click(confirmButton)

        await waitFor(() => container.getByText("Please try setting up MultiFactor Authentication after 30 minutes"))

        const title = container.getByText("setMFATitle");
        expect("setMFATitle").toEqual(title.textContent);
    });

    test("is back to login clicked", () => {
        const confirmButton = container.getByRole("button", { name: /backToLoginBtn/i });
        fireEvent.click(confirmButton)
    });
});

describe("Multi-Factor Authentication", () => {
    test("render with mfa cookie false", () => {
        const container = render(<MfaPage />)
    });
});
