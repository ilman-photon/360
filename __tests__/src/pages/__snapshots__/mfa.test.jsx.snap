// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Multi-Factor Authentication render with mfa cookie false 1`] = `
Object {
  "asFragment": [Function],
  "baseElement": <body>
    <div>
      <div
        class="container MuiBox-root css-0"
      >
        <div
          class="imageContainer MuiBox-root css-0"
        >
          <span
            style="box-sizing: border-box; display: inline-block; overflow: hidden; background: none; opacity: 1; border: 0px; margin: 0px; padding: 0px; position: relative; max-width: 100%;"
          >
            <span
              style="box-sizing: border-box; display: block; background: none; opacity: 1; border: 0px; margin: 0px; padding: 0px; max-width: 100%;"
            >
              <img
                alt=""
                aria-hidden="true"
                src="data:image/svg+xml,%3csvg%20xmlns=%27http://www.w3.org/2000/svg%27%20version=%271.1%27%20width=%27188%27%20height=%27188%27/%3e"
                style="display: block; max-width: 100%; background: none; opacity: 1; border: 0px; margin: 0px; padding: 0px;"
              />
            </span>
            <img
              alt="auth-image"
              data-nimg="intrinsic"
              decoding="async"
              src="data:image/gif;base64,R0lGODlhAQABAIAAAAAAAP///yH5BAEAAAAALAAAAAABAAEAAAIBRAA7"
              style="position: absolute; top: 0px; left: 0px; bottom: 0px; right: 0px; box-sizing: border-box; padding: 0px; margin: auto; display: block; width: 0px; height: 0px; min-width: 100%; max-width: 100%; min-height: 100%; max-height: 100%; object-fit: contain;"
            />
            <noscript />
          </span>
        </div>
        <div
          class="css-1p5q5e5-MuiStack-root"
        >
          <div
            class="contentContainer MuiBox-root css-0"
          >
            <h1
              class="MuiTypography-root MuiTypography-h1 title css-o2w69a-MuiTypography-root"
            >
              setMFATitle
            </h1>
            <p
              class="MuiTypography-root MuiTypography-body2 description css-e784if-MuiTypography-root"
            >
              setMFADescription
            </p>
            <p
              class="MuiTypography-root MuiTypography-body2 description css-e784if-MuiTypography-root"
            />
            <div
              class="checkBoxContainer MuiBox-root css-0"
            >
              <div
                class="checkBox MuiBox-root css-0"
              >
                <span
                  class="MuiCheckbox-root MuiCheckbox-colorPrimary MuiButtonBase-root MuiCheckbox-root MuiCheckbox-colorPrimary PrivateSwitchBase-root css-ua6y4i-MuiButtonBase-root-MuiCheckbox-root"
                >
                  <input
                    class="PrivateSwitchBase-input css-1m9pwf3"
                    data-indeterminate="false"
                    type="checkbox"
                  />
                  <svg
                    aria-hidden="true"
                    class="MuiSvgIcon-root MuiSvgIcon-fontSizeMedium css-i4bv87-MuiSvgIcon-root"
                    data-testid="CheckBoxOutlineBlankIcon"
                    focusable="false"
                    viewBox="0 0 24 24"
                  >
                    <path
                      d="M19 5v14H5V5h14m0-2H5c-1.1 0-2 .9-2 2v14c0 1.1.9 2 2 2h14c1.1 0 2-.9 2-2V5c0-1.1-.9-2-2-2z"
                    />
                  </svg>
                  <span
                    class="MuiTouchRipple-root css-8je8zh-MuiTouchRipple-root"
                  />
                </span>
                <p
                  class="MuiTypography-root MuiTypography-body1 checkBoxLabel css-ahj2mt-MuiTypography-root"
                >
                  rememberMeLabel
                </p>
              </div>
              <p
                class="MuiTypography-root MuiTypography-body2 checkBoxText css-e784if-MuiTypography-root"
              >
                rememberMeDescription
              </p>
            </div>
            <div
              class="css-1p5q5e5-MuiStack-root"
            >
              <button
                class="MuiButton-root MuiButton-contained MuiButton-containedPrimary MuiButton-sizeMedium MuiButton-containedSizeMedium MuiButtonBase-root customButton customButtonSmall  css-i2fzad-MuiButtonBase-root-MuiButton-root"
                data-testid="primary-button"
                tabindex="0"
                type="button"
              >
                confrimBtn
                <span
                  class="MuiTouchRipple-root css-8je8zh-MuiTouchRipple-root"
                />
              </button>
              <button
                class="MuiButton-root MuiButton-outlined MuiButton-outlinedPrimary MuiButton-sizeMedium MuiButton-outlinedSizeMedium MuiButtonBase-root customButton customButtonSmall  css-1yd4il8-MuiButtonBase-root-MuiButton-root"
                data-testid="secondary-button"
                tabindex="0"
                type="button"
              >
                backToLoginBtn
                <span
                  class="MuiTouchRipple-root css-8je8zh-MuiTouchRipple-root"
                />
              </button>
              <a
                class="MuiTypography-root MuiTypography-inherit MuiLink-root MuiLink-underlineAlways link css-1ps4owl-MuiTypography-root-MuiLink-root"
              />
            </div>
          </div>
        </div>
      </div>
    </div>
  </body>,
  "container": <div>
    <div
      class="container MuiBox-root css-0"
    >
      <div
        class="imageContainer MuiBox-root css-0"
      >
        <span
          style="box-sizing: border-box; display: inline-block; overflow: hidden; background: none; opacity: 1; border: 0px; margin: 0px; padding: 0px; position: relative; max-width: 100%;"
        >
          <span
            style="box-sizing: border-box; display: block; background: none; opacity: 1; border: 0px; margin: 0px; padding: 0px; max-width: 100%;"
          >
            <img
              alt=""
              aria-hidden="true"
              src="data:image/svg+xml,%3csvg%20xmlns=%27http://www.w3.org/2000/svg%27%20version=%271.1%27%20width=%27188%27%20height=%27188%27/%3e"
              style="display: block; max-width: 100%; background: none; opacity: 1; border: 0px; margin: 0px; padding: 0px;"
            />
          </span>
          <img
            alt="auth-image"
            data-nimg="intrinsic"
            decoding="async"
            src="data:image/gif;base64,R0lGODlhAQABAIAAAAAAAP///yH5BAEAAAAALAAAAAABAAEAAAIBRAA7"
            style="position: absolute; top: 0px; left: 0px; bottom: 0px; right: 0px; box-sizing: border-box; padding: 0px; margin: auto; display: block; width: 0px; height: 0px; min-width: 100%; max-width: 100%; min-height: 100%; max-height: 100%; object-fit: contain;"
          />
          <noscript />
        </span>
      </div>
      <div
        class="css-1p5q5e5-MuiStack-root"
      >
        <div
          class="contentContainer MuiBox-root css-0"
        >
          <h1
            class="MuiTypography-root MuiTypography-h1 title css-o2w69a-MuiTypography-root"
          >
            setMFATitle
          </h1>
          <p
            class="MuiTypography-root MuiTypography-body2 description css-e784if-MuiTypography-root"
          >
            setMFADescription
          </p>
          <p
            class="MuiTypography-root MuiTypography-body2 description css-e784if-MuiTypography-root"
          />
          <div
            class="checkBoxContainer MuiBox-root css-0"
          >
            <div
              class="checkBox MuiBox-root css-0"
            >
              <span
                class="MuiCheckbox-root MuiCheckbox-colorPrimary MuiButtonBase-root MuiCheckbox-root MuiCheckbox-colorPrimary PrivateSwitchBase-root css-ua6y4i-MuiButtonBase-root-MuiCheckbox-root"
              >
                <input
                  class="PrivateSwitchBase-input css-1m9pwf3"
                  data-indeterminate="false"
                  type="checkbox"
                />
                <svg
                  aria-hidden="true"
                  class="MuiSvgIcon-root MuiSvgIcon-fontSizeMedium css-i4bv87-MuiSvgIcon-root"
                  data-testid="CheckBoxOutlineBlankIcon"
                  focusable="false"
                  viewBox="0 0 24 24"
                >
                  <path
                    d="M19 5v14H5V5h14m0-2H5c-1.1 0-2 .9-2 2v14c0 1.1.9 2 2 2h14c1.1 0 2-.9 2-2V5c0-1.1-.9-2-2-2z"
                  />
                </svg>
                <span
                  class="MuiTouchRipple-root css-8je8zh-MuiTouchRipple-root"
                />
              </span>
              <p
                class="MuiTypography-root MuiTypography-body1 checkBoxLabel css-ahj2mt-MuiTypography-root"
              >
                rememberMeLabel
              </p>
            </div>
            <p
              class="MuiTypography-root MuiTypography-body2 checkBoxText css-e784if-MuiTypography-root"
            >
              rememberMeDescription
            </p>
          </div>
          <div
            class="css-1p5q5e5-MuiStack-root"
          >
            <button
              class="MuiButton-root MuiButton-contained MuiButton-containedPrimary MuiButton-sizeMedium MuiButton-containedSizeMedium MuiButtonBase-root customButton customButtonSmall  css-i2fzad-MuiButtonBase-root-MuiButton-root"
              data-testid="primary-button"
              tabindex="0"
              type="button"
            >
              confrimBtn
              <span
                class="MuiTouchRipple-root css-8je8zh-MuiTouchRipple-root"
              />
            </button>
            <button
              class="MuiButton-root MuiButton-outlined MuiButton-outlinedPrimary MuiButton-sizeMedium MuiButton-outlinedSizeMedium MuiButtonBase-root customButton customButtonSmall  css-1yd4il8-MuiButtonBase-root-MuiButton-root"
              data-testid="secondary-button"
              tabindex="0"
              type="button"
            >
              backToLoginBtn
              <span
                class="MuiTouchRipple-root css-8je8zh-MuiTouchRipple-root"
              />
            </button>
            <a
              class="MuiTypography-root MuiTypography-inherit MuiLink-root MuiLink-underlineAlways link css-1ps4owl-MuiTypography-root-MuiLink-root"
            />
          </div>
        </div>
      </div>
    </div>
  </div>,
  "debug": [Function],
  "findAllByAltText": [Function],
  "findAllByDisplayValue": [Function],
  "findAllByLabelText": [Function],
  "findAllByPlaceholderText": [Function],
  "findAllByRole": [Function],
  "findAllByTestId": [Function],
  "findAllByText": [Function],
  "findAllByTitle": [Function],
  "findByAltText": [Function],
  "findByDisplayValue": [Function],
  "findByLabelText": [Function],
  "findByPlaceholderText": [Function],
  "findByRole": [Function],
  "findByTestId": [Function],
  "findByText": [Function],
  "findByTitle": [Function],
  "getAllByAltText": [Function],
  "getAllByDisplayValue": [Function],
  "getAllByLabelText": [Function],
  "getAllByPlaceholderText": [Function],
  "getAllByRole": [Function],
  "getAllByTestId": [Function],
  "getAllByText": [Function],
  "getAllByTitle": [Function],
  "getByAltText": [Function],
  "getByDisplayValue": [Function],
  "getByLabelText": [Function],
  "getByPlaceholderText": [Function],
  "getByRole": [Function],
  "getByTestId": [Function],
  "getByText": [Function],
  "getByTitle": [Function],
  "queryAllByAltText": [Function],
  "queryAllByDisplayValue": [Function],
  "queryAllByLabelText": [Function],
  "queryAllByPlaceholderText": [Function],
  "queryAllByRole": [Function],
  "queryAllByTestId": [Function],
  "queryAllByText": [Function],
  "queryAllByTitle": [Function],
  "queryByAltText": [Function],
  "queryByDisplayValue": [Function],
  "queryByLabelText": [Function],
  "queryByPlaceholderText": [Function],
  "queryByRole": [Function],
  "queryByTestId": [Function],
  "queryByText": [Function],
  "queryByTitle": [Function],
  "rerender": [Function],
  "unmount": [Function],
}
`;
